src/utils/TrustedForwarder.sol

```diff
@@ -0,0 +1,120 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.20;
+
+import {ERC2771Context} from
+    "lib/openzeppelin-contracts/contracts/metatx/ERC2771Context.sol";
+
+/// @dev Minimal Trusted Multicall Forwarder
+/// @custom:derived from github.com/Synthetixio/trusted-multicall-forwarder
+/// @author Jared Borders <jaredborders@pm.me>
+contract TrustedForwarder {
+    struct Call {
+        /// @notice The address called
+        address target;
+        /// @notice the allocated value for the call defined by this struct
+        uint256 value;
+        /// @notice the call data for the call defined by this struct
+        bytes callData;
+    }
+
+    /// @dev thrown when requestedValue doesn't match with the available msgValue
+    error MismatchedValue(uint256 requestedValue, uint256 msgValue);
+
+    /// @dev thrown when the target contract does not trust this forwarder
+    error TargetNotTrusted(address target);
+
+    /// @notice Aggregate calls with a msg value
+    /// @notice Reverts if msg.value is less than the sum of the call values
+    /// @param calls An array of Call structs
+    /// @return returnData An array of results from the calls
+    function aggregate(Call[] calldata calls)
+        public
+        payable
+        returns (bytes[] memory returnData)
+    {
+        uint256 valAccumulator;
+        Call calldata calli;
+
+        uint256 length = calls.length;
+        returnData = new bytes[](length);
+
+        for (uint256 i = 0; i < length;) {
+            // define the call object
+            calli = calls[i];
+
+            /// @dev early exit;
+            /// ensure the target trusts this forwarder else revert the whole tx
+            if (!_isTrustedByTarget(calli.target)) {
```

Comment:
**Question:** Doesn't this severely limit the number of targets that can be called, most likely down to only one, being the SMv3 `Engine`? In which case, do we even need to allow users to be able to specify a target? It could send every call to the `Engine`? That would then limit the ability to execute arbitrary code to the `fulfillOracleQuery` and `withdrawEth` functions.